File: src/kernel/archs/x86_64/syscall.zig
const as = @import("./asm.zig");
const GdtType = @import("./gdt.zig").GdtType;
const syscall = @import("kernel").syscall;
const Registers = @import("./regs.zig").Registers;

const navy = @import("navy");

extern fn syscall_handle() void;

const Offset = struct {
    const star_kcs_off = 32;
    const star_ucs_off = 48;
};

pub fn setup() void {
    as.writeMsr(as.Msr.efer, as.readMsr(as.Msr.efer) | 1);

    var star: u64 = 0;
    star |= (@intFromEnum(GdtType.KernelCode) * 8) << Offset.star_kcs_off;
    star |= (((@intFromEnum(GdtType.UserCode) * 8) - 0x10) | 3) << Offset.star_ucs_off;

    as.writeMsr(as.Msr.star, star);
    as.writeMsr(as.Msr.lstar, @intFromPtr(&syscall_handle));
    as.writeMsr(as.Msr.syscall_flag_mask, 0xfffffffe);
}

pub fn set_gs(addr: u64) void {
    as.writeMsr(as.Msr.gs_base, addr);
    as.writeMsr(as.Msr.kern_gs_base, addr);
}

export fn syscall_handler(regs: *Registers) callconv(.C) void {
    regs.rax = syscall.handle(regs.rax, .{
        .arg1 = regs.rdi,
        .arg2 = regs.rsi,
        .arg3 = regs.rdx,
        .arg4 = regs.r8,
        .arg5 = regs.r9,
        .arg6 = regs.r10,
    });
}

File: src/kernel/archs/x86_64/syscall.c
#include <hal>
#include <logger>

#include "asm.h"
#include "gdt.h"
#include "syscall.h"

void syscall_init(void)
{
    asm_write_msr(MSR_EFER, asm_read_msr(MSR_EFER) | 1);
    asm_write_msr(MSR_STAR, ((uint64_t)(GDT_KERNEL_CODE * 8) << STAR_KCODE_OFFSET) | ((uint64_t)(((GDT_USER_DATA - 1) * 8) | 3) << STAR_UCODE_OFFSET));
    asm_write_msr(MSR_LSTAR, (uint64_t)syscall_handle);
    asm_write_msr(MSR_SYSCALL_FLAG_MASK, 0xfffffffe);
}

void syscall_set_gs(uintptr_t addr)
{
    asm_write_msr(MSR_GS_BASE, addr);
    asm_write_msr(MSR_KERN_GS_BASE, addr);
}

void syscall_handler(HalRegs *regs)
{
    regs->rax = _syscall_handler(regs->rax, (SysArgs){regs->rdi, regs->rsi, regs->rdx, regs->r8, regs->r9, regs->r10});
    return;
}
